src/general/semidiscretization.jl

```diff
@@ -286,6 +292,15 @@ function semidiscretize(semi, tspan; reset_threads=true)
         write_v0!(v0_system, system)
     end
 
+    if !isnothing(data_type)
+        # Convert all arrays to the correct array type. When e.g. `data_type=CuArray`, this will
+        # convert all `Array`s to `CuArray`s, moving data to the GPU.
+        # See the comments in general/gpu.jl for more details.
+        semi_adapted = Adapt.adapt(data_type, semi)
+
+        return DynamicalODEProblem(kick!, drift!, v0_ode, u0_ode, tspan, semi_adapted)
```

Comment:
can you not just overwrite `semi`?

Reply:
No, `Semidiscretization` is not mutable, and cannot be mutable to be GPU-compatible.